# Documentation for script: _findDistantFamilyMembers.pl_

## Summary
This program is used when we want to identify remote members of a family in TCDB in order to increase the coverage of sequence diversity within a family. It consists of BLASTing all members within a family against the NCBI NR database and selecting hits that show borderline similarity to other members in the family, while having no significant similarity to any other families in TCDB. Alignments should satisfy a minimal coverage and involve a minimal number of TMSs compatible with the topology of the family under analysis. This produces a list of candidates that should be further evaluated by an expert. 


## Contributor
Arturo Medrano-Soto


## Dependencies
The following dependencies need to be available in your enviroment for this 
program to run properly:

6. **_PERL 5.18_**  
Visit the [official website](https://www.perl.org/). This program 
was not tested with more recent versions of perl.

1. **Perl Module: _TCDB::CheckDependencies_**  
This module is included in the [TCDBtools distribution](https://github.com/SaierLaboratory/TCDBtools). 

2. **_Bioperl_**  
This module is included in the [TCDBtools distribution](https://github.com/SaierLaboratory/TCDBtools). 

3. **_HMMTOP_**  
This module is included in the [TCDBtools distribution](https://github.com/SaierLaboratory/TCDBtools). 

4. **_ssearch36 version: 36.3.8e_**  
Other versions of ssearch may require minor adaptations. Visit the
[download site](https://fasta.bioch.virginia.edu/fasta_www2/fasta_down.shtml). 



## Command line options
The following options are available. You can also run the 
script without arguments to display the options:


    -f, --family {string} (Optional)
      TCID of the family for which Pfam domain analysis will be carried out.
      If multiple TCIDs are given, they whould be comma-separated. The analysis
      will be performed individually for each family, unless the flag -sf is
      given, in which case all TCIDs will be treated as a superfamily.
      This option is incompatible with option -pt. But either -f or -pt
      must be given.

    -pt, --proj-targets {string}
      Project the characteristic domains of a reference family onto
      candidate protein(s) not in TCDB that might belong to the family.
      The format indicates pairs tcid,targets separated by ':'. That is:
      
      {tcid_1},{seq_file_1}:{tcid_2},{seq_file_2}:... {tcid_n},{seq_file_n}
      
      where tcid_n is the reference family that will project its domains onto
      the sequences in seq_file_n. This option is incompatible with option
      -f. But either -f or -pt must be given.
  
    -dc, --domain-cov {float} (Default: 0.7)
      Minimum coverage of the Pfam domain to consider it a match. If coverage
      is less than the specified threshold, the coverage must apply to the
      the query protein to consider the domain hit significant.

    -pc, --protein-cov {float} (Default: 0.1)
      Minimum coverage of the query protein sequence per domain.

    -e, --evalue {float} (Default: 1e-5)
      Maximum evalue threshold to consider a Pfam domain hit.

    -m, --prots-w-domain {float) (Default: 0.10)
      Minimum proportion of the proteins in the input family that should
      contain a domain, in order to consider the domain as part of the
      family for the purpose of this analyis.

    -o, --outdir {path} (Default: .)
      Directory where results and intermediary files will be saved.

    -s, --tcdb-seqs {file} (Mandatory)
      FASTA file with all sequences in TCDB (as generated by program
      extractFamily.pl). This file will be used to extract sequences
      and TCIDs for all members of the input family. This allows to
      freeze TCDB contents at a specific date, for the purpose defining
      a project.

    -sf, --superfamily {flag} (default: negated as --no-superfamily)
      This indicates that all families passed to option -f will be
      treated as a superfamily (i.e. as a single family) for the purpose of
      the Pfam domain  analysis. This is useful when a superfamily is
      composed of two or more different family TCIDs.

    -pfam {file} (Mandatory)
      The output of running hmmscan against all TCDB, or at least the input
      family(-ies).

    -b, --blastdb {path} (Default: assumed available in $BLASTDB)
      Full path to the blast database containing all the sequences in
      TCDB, this is to easily extract segments of proteins that match
      Pfam domains.

    -p, --rescue-prog (Default: ssearch36)
      Program that will be used to project or rescue domains in
      proteins that did have direct hits with Pfam domains.

    -h, --help
      Display this help. If present, this option takes precedence over any
      other option.
