# Documentation for script: _getDomainTopology.pl_

## Summary
This program characterizes the Pfam domain composition of a TCDB family. It first runs 
HMMSCAN on every protein sequence. If not all proteins in the family have a direct hit 
with a Pfam domain, the fragments of the sequences with direct hits are extracted and 
aligned to the sequences without direct hits using the Smith-Waterman algorithm as 
implemented in SSEARCH. This is equivalent to projecting the domains of sequences with 
direct PFAM hits onto sequences without Pfam hits. The expectation is that all sequences 
in a family should share at least one domain or have domains that belong to the same 
Pfam clan. This can further be used to relate families based on the overlap of shared 
domains. 

## How to cite this program
The rational for this program was published in two of our publications: 
(PMID: 29145176 and 29579047):

  * Medrano-Soto A, Moreno-Hagelsieb G, McLaughlin D, Ye ZS, Hendargo KJ, Saier MH Jr. 
  _Bioinformatic characterization of the Anoctamin Superfamily of Ca2+-activated ion 
  channels and lipid scramblases._  2018. PLoS One. **13**(3):e0192851
  **PMID:** [29579047](https://www.ncbi.nlm.nih.gov/pubmed/?term=29579047) 
  
  * Moreno-Hagelsieb G, Vitug B, Medrano-Soto A, Saier MH Jr.
  J Mol Microbiol Biotechnol. 2017;27(4):252-267
  **PMID:** [29145176](https://www.ncbi.nlm.nih.gov/pubmed/?term=29145176)


## Dependencies
The following programs need to be available in your path for this 
program to run properly:

1. **_extractFamily.pl_**  
This program is available in our [Perl repository](https://github.com/SaierLaboratory/TCDBtools). 

2. **_famXpander.pl_**  
This program is available in our [Perl repository](https://github.com/SaierLaboratory/TCDBtools). 

3. **_protocol2.py_**  
This program is available in our [Python repository](https://github.com/SaierLaboratory/BioVx).

4. **_gsat.py_**  
This program is available in our [Python repository](https://github.com/SaierLaboratory/BioVx).

5. **_PERL 5.18_**  
Visit the [official website](https://www.perl.org/). This program 
was not tested with more recent versions of perl.

## Command line options
The following options are available. You can also run the 
script without arguments to display the options:


 -f, --family {string} (Mandatory)
  TCID of the family for which Pfam domain analysis will be carried out.
  If multiple TCIDs are given, they whould be comma-separated. The analysis
  will be performed individually for each family, unless the flag -sf is
  given, in which case all TCIDs will be treated as a superfamily.

 -dc, --domain-cov {float} (Default: 0.7)
  Minimum coverage of the Pfam domain to consider it a match. If coverage
  is less than the specified threshold, the coverage must apply to the
  the query protein to consider the domain hit significant.

 -pc, --protein-cov {float} (Default: 0.1)
  Minimum coverage of the query protein sequence per domain.

 -e, --evalue {float} (Default: 1e-5)
  Maximum evalue threshold to consider a Pfam domain hit.

 -m, --prots-w-domain {float) (Default: 0.10)
  Minimum proportion of the proteins in the input family that should
  contain a domain, in order to consider the domain as part of the
  family for the purpose of this analyis.

 -o, --outdir {path} (Default: .)
  Directory where results and intermediary files will be saved.

 -s, --tcdb-seqs {file} (Mandatory)
  FASTA file with all sequences in TCDB (as generated by program
  extractFamily.pl). This file will be used to extract sequences
  and TCIDs for all members of the input family. This allows to
  freeze TCDB contents at a specific date, for the purpose defining
  a project.

 -sf, --superfamily {flag} (default: negated as --no-superfamily)
  This indicates that all families passed to option -f will be
  treated as a superfamily (i.e. as a single family) for the purpose of
  the Pfam domain  analysis. This is useful when a superfamily is
  composed of two or more different family TCIDs.

 -pfam {file} (Mandatory)
  The output of running hmmscan against all TCDB, or at least the input
  family(-ies).

 -b, --blastdb {path} (Default: assumed available in $BLASTDB)
  Full path to the blast database containing all the sequences in
  TCDB, this is to easily extract segments of proteins that match
  Pfam domains.

 -p, --rescue-prog (Default: ssearch36)
  Program that will be used to project or rescue domains in
  proteins that did have direct hits with Pfam domains.

 -h, --help
  Display this help. If present, this option takes precedence over any
  other option.
